Detection Transformer Estimator

kieranfraser opened the pull request:
# Description

Adding **DE**tection **TR**ansformer (DETR) to ART to facilitate object detection using Transformers.

- DETR estimator added to art/estimators/object_detection
- requires additional method in art/estimators/pytorch.py which sets MultiheadAttention layer into train/eval mode
- associated tests added
- notebook demonstrating use of DETR estimator detecting objects and under adversarial patch attack added to notebooks/adversarial_patch/ and updated notebooks/ README

Paper: ["End-to-End Object Detection with Transformers"](https://arxiv.org/abs/2005.12872) 

Fixes # (issue)
#1638

## Type of change

Please check all relevant options.

- [x] Improvement (non-breaking)
- [ ] Bug fix (non-breaking)
- [x] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

# Testing

Please describe the tests that you ran to verify your changes. Consider listing any relevant details of your test configuration.

- [x] Test prediction of DETR
- [x] Test loss gradient
- [x] Test applying preprocessing defences
- [x] Test computation of losses (classification, bbox, giou)
- [x] Test DETR with PGD attack

**Test Configuration**:
- OS
- Python version
- ART version or commit number
- TensorFlow / Keras / PyTorch / MXNet version

# Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

beat-buesser commented on art/estimators/object_detection/pytorch_detection_transformer.py:
```diff
@@ -0,0 +1,998 @@
+logger = logging.getLogger(__name__)
+
+
+def box_cxcywh_to_xyxy(x: "torch.Tensor"):
```
We need to mention the license of these functions in their docstrings.

beat-buesser commented on art/estimators/object_detection/pytorch_detection_transformer.py:
```diff
@@ -0,0 +1,998 @@
+logger = logging.getLogger(__name__)
+
+
+def box_cxcywh_to_xyxy(x: "torch.Tensor"):
```
We need to mention the license of these functions in their docstrings.

kieranfraser replied:
Added reference to Apache 2.0

beat-buesser commented on art/estimators/object_detection/pytorch_detection_transformer.py:
```diff
@@ -0,0 +1,998 @@
+        """
+        import torch
+
+        class HungarianMatcher(torch.nn.Module):
```
We need to mention the license of these functions in their docstrings.

kieranfraser replied:
Added reference to Apache 2.0

beat-buesser commented on art/estimators/object_detection/pytorch_detection_transformer.py:
```diff
@@ -0,0 +1,998 @@
+        return x, targets
+
+
+class NestedTensor:
```
We need to mention the license of these functions in their docstrings.

kieranfraser replied:
Added reference to Apache 2.0

beat-buesser reviewed (changes requested):
Hi @kieranfraser Thank you very much for your PR. I think we need to include the license of reused code somewhere. I think the patches don't have a lot of effects on the detections, is there a way how we could generate a stronger patch?

kieranfraser commented:
> Hi @kieranfraser Thank you very much for your PR. I think we need to include the license of reused code somewhere. I think the patches don't have a lot of effects on the detections, is there a way how we could generate a stronger patch?

Thanks @beat-buesser. I have added references to the license (Apache 2.0) in the `Docstrings` of reused code. Also updated the notebook with stronger Adversarial Patch attacks (targeted and untargeted). Seems that unless a fixed patch location is passed, the attack fails to converge to a patch that fools the DETR - plan to experiment further with this. But the notebook now demonstrates working patches as well as PGD for the DETR.
